fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/LoanApplicationTerms.java

```diff
@@ -223,10 +224,180 @@ public final class LoanApplicationTerms {
+        }
+
+        public Builder repaymentsStartingFromDate(LocalDate repaymentsStartingFromDate) {
+            this.repaymentsStartingFromDate = repaymentsStartingFromDate;
+            return this;
+        }
+
+        public Builder daysInMonthType(DaysInMonthType daysInMonthType) {
+            this.daysInMonthType = daysInMonthType;
+            return this;
+        }
+
+        public Builder daysInYearType(DaysInYearType daysInYearType) {
+            this.daysInYearType = daysInYearType;
+            return this;
+        }
+
+        public Builder variationsDataWrapper(LoanTermVariationsDataWrapper variationsDataWrapper) {
+            this.variationsDataWrapper = variationsDataWrapper;
+            return this;
+        }
+
+        public Builder fixedLength(Integer fixedLength) {
+            this.fixedLength = fixedLength;
+            return this;
+        }
+
+        public Builder inArrearsTolerance(Money inArrearsTolerance) {
+            this.inArrearsTolerance = inArrearsTolerance;
+            return this;
+        }
+
+        public Builder disbursementDatas(List<DisbursementData> disbursementDatas) {
+            this.disbursementDatas = disbursementDatas;
+            return this;
+        }
+
+        public Builder downPaymentAmount(Money downPaymentAmount) {
+            this.downPaymentAmount = downPaymentAmount;
+            return this;
+        }
+
+        public LoanApplicationTerms build() {
+            return new LoanApplicationTerms(this);
+        }
+    }
+
+    public static LoanApplicationTerms assembleFrom(LoanRepaymentScheduleModelData modelData) {
+        Money principal = Money.of(modelData.currency().toData(), modelData.disbursementAmount());
+        Money downPaymentAmount = Money.zero(modelData.currency().toData());
+
+        if (modelData.downPaymentEnabled()) {
+            downPaymentAmount = Money.of(modelData.currency().toData(),
+                    MathUtil.percentageOf(principal.getAmount(), modelData.disbursementAmount(), 19));
+            if (modelData.installmentAmountInMultiplesOf() != null) {
+                downPaymentAmount = Money.roundToMultiplesOf(downPaymentAmount, modelData.installmentAmountInMultiplesOf());
+            }
+        }
+
+        return new Builder().currency(modelData.currency()).loanTermFrequency(modelData.numberOfRepayments())
+                .loanTermPeriodFrequencyType(PeriodFrequencyType.valueOf(modelData.repaymentFrequencyType()))
+                .numberOfRepayments(modelData.numberOfRepayments()).repaymentEvery(modelData.repaymentFrequency())
+                .repaymentPeriodFrequencyType(PeriodFrequencyType.valueOf(modelData.repaymentFrequencyType()))
+                .interestRatePerPeriod(modelData.nominalInterestRate())
+                .interestRatePeriodFrequencyType(PeriodFrequencyType.valueOf(modelData.repaymentFrequencyType()))
+                .annualNominalInterestRate(modelData.nominalInterestRate()).principal(principal)
```

Comment:
Can we just use a better field name for the incoming `nominalInterestRate` parameter?
As I understood this field holds yearly Interest rate in this case.

Reply:
Changed to annualNominalInterestRate